### PR TITLE
New DynamicProperty that is created if a Property is used as a wrapper

### PR DESCRIPTION
**Pull request by fwkoch**
See: #105

### CODE REVIEW

File: properties/basic.py [hunk 1]
```diff
@@ -231,6 +230,80 @@ def sphinx_class(self):
...
+
+    def __init__(self, doc, func, prop, **kwargs):
+        self.func = func
+        self.prop = prop
+        self.name = func.__name__
+        super(DynamicProperty, self).__init__(doc, **kwargs)
+
+    @property
+    def func(self):
+        """func is used do calculate the dynamic value"""
```
Comment (rowanc1) [inline properties/basic.py]: func is used **do** calculate the dynamic value

File: properties/basic.py [hunk 2]
```diff
@@ -231,6 +230,80 @@ def sphinx_class(self):
             pref=self.__module__
         )
 
+    def __call__(self, func):
+        return DynamicProperty(self.doc, func=func, prop=self)
```
Comment (rowanc1) [inline properties/basic.py]: clever, just wrapping a new class.

**Comment by rowanc1** [comment]
I guess there was some discussion in #105 that we didn't want to support setters. However, the location example is actually kind of a good one. You might want a vector property that is actually defined as three independent floats.

```python
props.x = 2
print(props.loc)
# [2, 0, 0]
props.loc = [4, 5]
# raises error
props.loc = [4, 5, 6]
print(props.x)
# 4
```

Thoughts?

**Comment by fwkoch** [comment]
Yeah, ok, I really like that after thinking about it a bit more.

`DynamicProperties` are calculated dynamically from other properties, can be used to set other properties. They are never saved to the `_backend` (no independent info to save, can always be recalculated if you have the `HasProperties` class), they are not serialized (unless custom serialization is used), they do not fire change notifications. `validator` functionality can just be baked into the wrapped methods; `observer` functionality should just be on the properties that they are used to set. They do not have `default` value (I don't think?) - those should be baked into the wrapped methods too.

**Comment by bsmithyman** [comment]
@fwkoch: Right, so they are purely functional (like mappings, etc.). They have no state, and so they don't actually affect the data model, but are purely an interface thing. Sounds cool :)

**Comment by fwkoch** [comment]
@bsmithyman - exactly. Basically anywhere that you would rather just use `@property` rather than make a new `Property`, but you still want documentation and validation.

**Comment by fwkoch** [comment]
Tests here are failing because of pylint - unhappy that files are so long. I agree, so I fixed it in a separate PR: https://github.com/3ptscience/properties/pull/109

**Comment by fwkoch** [comment]
Notebook snippet to show `DynamicProperty` in action:

screen shot 2016-12-12 at 1 55 47 pm

**Comment by varyag00** [comment]
Not much in term of feedback, and the code I wrote using it didn't end up going to production, but from the basic use I did get out of it I found it handy. It worked exactly as expected with no surprises or hiccups  (i.e. properties as a decorator).

Pycharm didn't always play nicely with it in terms of autocomplete and highlighting it as a syntax error, but I'm thinking that was due to me using multiple conda envs with different versions of properties (some of them not on this branch) throughout the process without restarting it. Maybe it got confused? 

Overall super cool, looking forward to getting more use out of it.

**Pull request merged**